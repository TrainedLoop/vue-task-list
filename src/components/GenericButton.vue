<template>
    <div :class="buttonClasses">
        <button v-bind="$attrs" type="button" @click="onClick">
            {{ props.text }}
        </button>
    </div>
</template>

<script setup lang='ts'>

export interface Props {
    onClick: () => void;
    type?: 'success' | 'default';
    text?: string;
}

const props = withDefaults(defineProps<Props>(), {
    type: 'default',
});

const buttonClasses = ['generic-button']

switch (props.type) {
    case 'success':
        buttonClasses.push('generic-button--success')
        break;
    default:
        break;
}


</script>



<style lang="scss" scoped>
.generic-button {
    margin-top: 20px;

    & button {
        outline: none;
        background-color: $default-color-1;
        color: $text-color-light;
        height: 2.5rem;
        font-size: 1rem;
        border: none;
        box-shadow: inset 1px 2px 5px 2px rgba(0, 0, 0, 0),
            1px 2px 5px 2px rgba(0, 0, 0, 0.15);
        border-radius: 0.3rem;
        width: 100%;
        text-transform: uppercase;
        font-weight: bolder;
        transition: all .3s ease;
        transform: scale(1);

        &:active {
            box-shadow: inset 1px 2px 5px 2px rgba(0, 0, 0, 0.1),
                1px 2px 5px 2px rgba(0, 0, 0, 0);
            transition: box-shadow transform .1s ease;
        }

        &:hover {
            transform: scale(1.002);
            transition: box-shadow transform .3s ease;

        }

    }
    &--success {
        & button {
            background-color: $light-color-2;
            color: $text-color-light;
        }
    }
}
</style>