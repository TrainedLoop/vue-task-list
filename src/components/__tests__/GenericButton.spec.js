import { shallowMount, mount } from '@vue/test-utils';
import GenericButton from '../GenericButton.vue';

//https://testdriven.io/blog/vue-unit-testing/
describe('GenericInput.vue Tests', () => {
  let wrapper = null;
  let props;
  let onClick = jest.fn();
  beforeEach(() => {
    props = {
      text: 'Button Text',
      onClick: onClick,
    };
  });
  afterEach(() => {
    wrapper.unmount();
    jest.clearAllMocks();
  });

  it('component contains correct props', () => {
    wrapper = shallowMount(GenericButton, {
      propsData: props,
    });
    expect(wrapper.props()).toStrictEqual({
      text: 'Button Text',
      onClick: expect.any(Function),
      type: 'default',
    });
  });

  it('renders component correctly', () => {
    wrapper = shallowMount(GenericButton, {
      propsData: props,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('if button will be clicked', () => {
    it('should call onClick callback 1 time', () => {
      wrapper = shallowMount(GenericButton, {
        propsData: props,
      });
      wrapper.find('button').trigger('click');
      expect(onClick).toBeCalledTimes(1);
    });
  });

  describe('if type prop is "success"', () => {
    beforeEach(() => {
      props.type = 'success';
    });
    it('should component contains correct props', () => {
      wrapper = shallowMount(GenericButton, {
        propsData: props,
      });
      expect(wrapper.props()).toStrictEqual({
        text: 'Button Text',
        onClick: expect.any(Function),
        type: 'success',
      });
    });
    it('should contains correct variant class', async () => {
      wrapper = shallowMount(GenericButton, {
        propsData: props,
      });
      expect(wrapper.classes()).toContain('generic-button--success');
    });
  });
});
