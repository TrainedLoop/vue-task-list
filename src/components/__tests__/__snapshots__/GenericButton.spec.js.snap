// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericInput.vue Tests renders component correctly 1`] = `
<div
  class="generic-button"
>
  <button
    type="button"
  >
    Button Text
  </button>
</div>
`;
